### ISSUE TITLE
ESIL for x86 test sets SF incorrectly for non-64 bit

### ISSUE THREAD

**Issue by thestr4ng3r** [radareorg/radare2#14249]
Example:
```
e asm.arch=x86
e asm.bits=64
s 0
aei
aepc 0
wa test eax, eax
aer eax=-1
aes
aer all~sf
```
gives:
```
sf = 0x00000000
```
which is wrong. It does work however when using rax instead.

This is the esil for `test eax, eax`:
```
0,eax,eax,&,==,$z,zf,:=,$p,pf,:=,$s,sf,:=,0,cf,:=,0,of,:=
```
s eax is &-ed with itself and the result is compared against 0 to get the flags. The issue is that the `==` op at this point has no idea about the bitness and falls back to 64:
```c
// This function also sets internal vars which is used in flag calculations.
static int esil_cmp(RAnalEsil *esil) {
	ut64 num, num2;
	int ret = 0;
	char *dst = r_anal_esil_pop (esil);
	char *src = r_anal_esil_pop (esil);
	if (dst && r_anal_esil_get_parm (esil, dst, &num)) {
		if (src && r_anal_esil_get_parm (esil, src, &num2)) {
			esil->old = num;
			esil->cur = num - num2;
			ret = 1;
			if (r_reg_get (esil->anal->reg, dst, -1)) {
				esil->lastsz = esil_internal_sizeof_reg (esil, dst);
			} else if (r_reg_get (esil->anal->reg, src, -1)) {
				esil->lastsz = esil_internal_sizeof_reg (esil, src);
			} else {
				// default size is set to 64 as internally operands are ut64
				esil->lastsz = 64;
			}
		}
	}
	free (dst);
	free (src);
	return ret;
}
```
Imo, the fallback should be `esil->anal->bits` instead, but that won't solve `test eax, eax` in x86_64.

So to fix `test`, either one could pull out the sign bit manually from the result of the & and store it in sf or make == aware about the bitness, for example by adding ops like `==8` (lol), `==16`, `==32` and `==64`. I'd prefer to go with the second option, because if esil has a notion of signedness, it must also be explicit about how many bits it uses. This might also count some other ops apart from `==`.

**Comment by bacher09** [comment]
Hi, I just faced with this issue myself.  Here's simple C program that affected by this.

```
#include <stdio.h>

static __attribute__((used)) int g = 0;

void test() {
    int c = 3;
    asm volatile(
        "start%=:\n\t"
        "cmp $0, %0\t\n"
        "jl done%=\t\n"
        "mov %0, %%eax\t\n"
        "add %%eax, g\t\n"
        "sub $1, %%eax\t\n"
        "mov %%eax, %0\t\n"
        "jmp start%=\t\n"
        "done%=:"
        :
        : "o"(c)
        : "memory", "cc"
        );
}


int main() {
    test();
    printf("%d\n", g);
    return 0;
}

```

Executing `test` procedure with ESIL will take indefinitely.

**Comment by thestr4ng3r** [comment]
test has been fixed by 21fb73d1f8995cd7434524ec0f8778927d1b26aa
